### DISCUSSION TITLE
How can I add downstream gRPC support while keeping upstream communication over standard HTTP?

### DISCUSSION THREAD

**Discussion by simboyd** [ThreeMammals/Ocelot#1803] [Q&A] [answered]
Hi, I'd like to receive HTTPS requests from the client in my Ocelot gateway API and call existing grpc microservices to aggregate data and return it back to the client over HTTPS in JSON format. sorry if this is a silly question but how would I go about implementing a mechanism for translating HTTP requests into grpc messages downstream? Where do I begin?

**Comment by ks1990cn** [comment]
No you can not, it due to different protocols, GRPC have concurrency limit, you need different type of Load balancers. Probably need to write logic for socket so continuously it will be alive.
GRPC is made for high performance and you may loose that.
Try using it as per recommendations, please go for Rest API if you need highly scalled and highly perfomant system with JSON result back.
- [Performance best practices with gRPC](https://learn.microsoft.com/en-us/aspnet/core/grpc/performance)

**Reply by raman-m** [reply]
True concerning performance, but the author wants to integrate 😄

**Comment by raman-m** [comment, marked as answer]
> @simboyd [on Nov 25, 2022](#discussion-5879774)

Hello, Sandro!
Welcome to Ocelot world!
Sorry, we've missed the thing for a year...
Are you still with Ocelot?

----
> how would I go about implementing a mechanism for translating HTTP requests into GRPC messages downstream?

In theory it is possible. But that will be your custom solution, you see?

----
> Where do I begin?

As a known fact, Ocelot **does not** support GRPC protocol & scheme now. We plan to discuss GRPC in 2-3 months (in 2024). Good point to create a milestone...

There is only one option...
You need to develop a custom [Delegating Handler](https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html) to convert HTTP request to GRPC one. After that you have to process GRPC response to convert it back to HTTP one. So, delegating handlers feature is a great ASP.NET Core feature which helps to override everything! 😍 

That will be very interesting coding challenge! 😉 
Please, create Ocelot fork in [your repos](https://github.com/simboyd?tab=repositories)... And let's get started!

**Reply by Liliya-Nuraly** [reply]
Good afternoon! Does Ocelot still not support the GRPC protocol and scheme at this point?

**Reply by raman-m** [reply]
@Liliya-Nuraly No, unfortunately not!
However, we do have long-term plans to include gRPC in our development roadmap, but after closing of all current [25 pull requests](https://github.com/ThreeMammals/Ocelot/pulls).
Ocelot can route HTTP/2 traffic, and since gRPC is built on the HTTP/2 protocol, it aligns well. However, converting schemes from `http` to `grpc` can be a tricky task. Are there any proxies or gateways that handle this type of scheme conversion?
What are your interests and options when it comes to gRPC? Do you have intention to contribute?
